w3af/plugins/crawl/wpscan.py

```diff
@@ -0,0 +1,232 @@
+
+w3af is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation version 2 of the License.
+
+w3af is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+import json
+import os
+import urllib
+import zipfile
+
+from itertools import repeat, izip
+
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.constants.severity as severity
+import w3af.core.data.kb.knowledge_base as kb
+
+from w3af import ROOT_PATH
+
+from w3af.core.controllers.plugins.crawl_plugin import CrawlPlugin
+from w3af.core.controllers.exceptions import RunOnce
+from w3af.core.controllers.core_helpers.fingerprint_404 import is_404
+
+from w3af.core.data.options.opt_factory import opt_factory
+from w3af.core.data.options.option_list import OptionList
+from w3af.core.data.options.option_types import BOOL
+from w3af.core.data.fuzzer.utils import rand_alnum
+from w3af.core.data.db.disk_set import DiskSet
+from w3af.core.data.request.fuzzable_request import FuzzableRequest
+from w3af.core.data.kb.info import Info
+from w3af.core.data.kb.info_set import InfoSet
+from w3af.core.data.kb.vuln import Vuln
+from w3af.core.controllers.misc.decorators import runonce
+
+class wpscan(CrawlPlugin):
+    """
+    Finds WordPress plugins by bruteforcing.
+    
+    :author: jose nazario (jose@monkey.org)
+    """
+
+    BASE_PATH = os.path.join(ROOT_PATH, 'plugins', 'crawl', 'wpscan')
+
+    def __init__(self):
+        CrawlPlugin.__init__(self)
+        self._update_plugins = False
+        self._plugin_list = {}
+        # Internal variables
+        self._exec = True
+        self._already_tested = DiskSet(table_prefix='wpscan')
+
+    @runonce(exc_class=RunOnce)
+    def crawl(self, fuzzable_request):
+        """
+        Get the file and parse it.
+        
+        :param fuzzable_request: A fuzzable_request instance that contains
+                               (among other things) the URL to test.
+        """
+        if self._update_plugins or not os.path.exists(os.path.join(self.BASE_PATH, 'plugins.json')):
+            om.out.debug("Downloading the archive from wpscan and extracting plugins.json ...", )
+            try:
+                os.stat(BASE_PATH)
+            except OSError:
+                om.out.debug("%s doesn't exist, making ..." % BASE_PATH)
+                os.mkdir(BASE_PATH)
+                om.out.debug("done")
+            with open('/tmp/data.zip', 'wb') as f:
+                f.write(urllib.urlopen('https://github.com/wpscanteam/wpscan/raw/master/data.zip').read())
```

Comment:
Not so important: This wouldn't work in an environment with a proxy config, even if the user configured the proxy in w3af.

Rather important: This is not the usual way in which we do things in w3af. I prefer including the `plugins.json` file in our repository and have everything in one place ready for the user to download and use. What is the license for `plugins.json`?

There are other plugins, like phishtank for example, which have a unittest which checks the date of the data file and fails if the date is too old. This makes sure that we update the data file for the unittest to PASS -> for the build to PASS.